src/pointfree/mapAll.js

```diff
@@ -0,0 +1,21 @@
+/** @license ISC License (c) copyright 2016 original and current authors */
+/** @author Karthik Iyengar (karthikiyengar) */
+
+const isFunction = require('../core/isFunction')
+
+function mapAll() {
+  const parts = [].slice.call(arguments)
+  parts.forEach(fn => {
+    if (!isFunction(fn)) {
+      throw new TypeError('mapAll: All arguments must be Functions')
+    }
+  })
+  return (m) => {
+    if(!(m && isFunction(m.mapAll))) {
+      throw new TypeError('mapAll: Tuple required')
+    }
+    return m.mapAll(...parts)
+  }
+}
+
+module.exports = mapAll
```

Comment:
We should make this curry-able and would make more sense living in `src/Tuple` as it only works on tuple types.
If we rename this to `nMap` the implementation should look like this:

```js
const validTuple = (n, m) =>
  isSameType(Tuple(n), m)

function runMap(m, fns) {
  const n = fns.length

  if (!validTuple(n, m)) {
    throw new TypeError(`nMap: ${n}-Tuple required`)
  }

  fns.forEach(fn => {
    if (!isFunction(fn)) {
      throw new TypeError('nMap: Functions required')
    }
  })

  return m.mapAll(...fns)
}

function dispatch(n) {
  switch (n) {
  case 1:
    return (a, m) =>
      runMap(m, [ a ])
  case 2:
    return (a, b, m) =>
      runMap(m, [ a, b ])
  case 3:
    return (a, b, c, m) =>
      runMap(m, [ a, b, c ])
  case 4:
    return (a, b, c, d, m) =>
      runMap(m, [ a, b, c, d ])
  case 5:
    return (a, b, c, d, e, m) =>
      runMap(m, [ a, b, c, d, e ])
  case 6:
    return (a, b, c, d, e, f, m) =>
      runMap(m, [ a, b, c, d, e, f ])
  case 7:
    return (a, b, c, d, e, f, g, m) =>
      runMap(m, [ a, b, c, d, e, f, g ])
  case 8:
    return (a, b, c, d, e, f, g, h, m) =>
      runMap(m, [ a, b, c, d, e, f, g, h ])
  case 9:
    return (a, b, c, d, e, f, g, h, i, m) =>
      runMap(m, [ a, b, c, d, e, f, g, h, i ])
  case 10:
    return (a, b, c, d, e, f, g, h, i, j, m) =>
      runMap(m, [ a, b, c, d, e, f, g, h, i, j ])
  }
}

function nMap(n) {
  if (!isInteger(n) && (n < 1 || n > 10)) {
    throw new TypeError(
      'nMap: Integer between 1 and 10 required for first argument'
    )
  }

  return dispatch(n)
}

module.exports =
  curry(nMap)
```

It is a doozy, but we allows for currying, and all the partial application goodness that comes along with it.

Reply:
@evilsoft - I will wait for our decision regarding getting rid of the `switch/case` before I go ahead and fix this. I think it would make an order of magnitude difference in getting rid of the indirection and improving readability.